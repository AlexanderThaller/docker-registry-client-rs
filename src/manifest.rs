use serde::Deserialize;
use url::Url;

#[derive(Debug, Deserialize)]
pub(super) struct Image {
    #[serde(rename = "schemaVersion")]
    schema_version: SchemaVersion,

    #[serde(rename = "mediaType")]
    media_type: String,

    config: Config,

    layers: Vec<Layer>,
}

#[derive(Debug, Deserialize)]
pub(super) struct List {
    #[serde(rename = "schemaVersion")]
    schema_version: SchemaVersion,

    #[serde(rename = "mediaType")]
    media_type: String,

    manifests: Vec<Manifest>,
}

#[derive(Debug)]
pub(super) enum SchemaVersion {
    V2,
}

#[derive(Debug, Deserialize)]
pub(super) struct Manifest {
    #[serde(rename = "mediaType")]
    media_type: String,
    size: u64,
    digest: String,
    platform: Platform,
}

#[derive(Debug, Deserialize)]
pub(super) struct Platform {
    architecture: Architecture,
    os: OperatingSystem,

    #[serde(rename = "os.version")]
    os_version: Option<String>,

    #[serde(rename = "os.features")]
    os_features: Option<String>,

    variant: Option<String>,

    #[serde(default)]
    features: Vec<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "lowercase")]
pub(super) enum Architecture {
    #[serde(rename = "386")]
    I386,

    Amd64,
    Arm,
    Arm64,
    Loong64,
    Mips,
    Mips64,
    Mips64le,
    Mipsle,
    Ppc64,
    Ppc64le,
    Riscv64,
    S390x,
    Wasm,

    Unknown,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "lowercase")]
pub(super) enum OperatingSystem {
    Aix,
    Android,
    Darwin,
    Dragonfly,
    Freebsd,
    Illumos,
    Ios,
    Js,
    Linux,
    Netbsd,
    Openbsd,
    Plan9,
    Solaris,
    Wasip1,
    Windows,

    Unknown,
}

#[derive(Debug, Deserialize)]
pub(super) struct Config {
    #[serde(rename = "mediaType")]
    media_type: String,
    size: u64,
    digest: String,
}

#[derive(Debug, Deserialize)]
pub(super) struct Layer {
    #[serde(rename = "mediaType")]
    media_type: String,
    size: u64,
    digest: String,

    #[serde(default)]
    urls: Vec<Url>,
}

impl<'de> Deserialize<'de> for SchemaVersion {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let value = i32::deserialize(deserializer)?;

        match value {
            2 => Ok(Self::V2),
            _ => Err(serde::de::Error::custom("invalid enum value")),
        }
    }
}

#[cfg(test)]
mod tests {
    mod list {
        mod deserialize {
            use crate::manifest::List;

            #[test]
            fn example() {
                const INPUT: &str = include_str!("../resources/manifest/list/example.json");

                let out: List = serde_json::from_str(INPUT).unwrap();

                insta::assert_debug_snapshot!(out);
            }

            #[test]
            fn trivy() {
                const INPUT: &str = include_str!("../resources/manifest/list/trivy.json");

                let out: List = serde_json::from_str(INPUT).unwrap();

                insta::assert_debug_snapshot!(out);
            }
        }
    }

    mod image {
        mod deserialize {
            use crate::manifest::Image;

            #[test]
            fn example() {
                const INPUT: &str = include_str!("../resources/manifest/image/example.json");

                let out: Image = serde_json::from_str(INPUT).unwrap();

                insta::assert_debug_snapshot!(out);
            }
        }
    }
}
