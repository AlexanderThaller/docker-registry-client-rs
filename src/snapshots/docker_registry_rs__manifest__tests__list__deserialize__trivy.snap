---
source: src/manifest.rs
expression: out
---
List {
    schema_version: V2,
    media_type: "application/vnd.docker.distribution.manifest.list.v2+json",
    manifests: [
        Manifest {
            media_type: "application/vnd.docker.distribution.manifest.v2+json",
            size: 1159,
            digest: "sha256:4704989dd70bd0145e3820b6ce68cbfcc9a5e6e9a222a88ceaef1001dcccb1de",
            platform: Platform {
                architecture: Amd64,
                os: Linux,
                os_version: None,
                os_features: None,
                variant: None,
                features: [],
            },
        },
        Manifest {
            media_type: "application/vnd.docker.distribution.manifest.v2+json",
            size: 1159,
            digest: "sha256:c28826c9944b53ec9405bfd0efcf78a096e0970f38e4a2f0cdc62ea3fa0ea61e",
            platform: Platform {
                architecture: Arm64,
                os: Linux,
                os_version: None,
                os_features: None,
                variant: None,
                features: [],
            },
        },
        Manifest {
            media_type: "application/vnd.docker.distribution.manifest.v2+json",
            size: 1159,
            digest: "sha256:fd48d0f733fbf19f6ad8c6238330c163c64089f2c7d22f17d841287b456c087f",
            platform: Platform {
                architecture: Ppc64le,
                os: Linux,
                os_version: None,
                os_features: None,
                variant: None,
                features: [],
            },
        },
        Manifest {
            media_type: "application/vnd.docker.distribution.manifest.v2+json",
            size: 1159,
            digest: "sha256:289f91dc4759e9376f8124715363b33a282fc7c704be6aa7e3852b966c40c084",
            platform: Platform {
                architecture: S390x,
                os: Linux,
                os_version: None,
                os_features: None,
                variant: None,
                features: [],
            },
        },
    ],
}
